cmake/detect_gpu.cmake

```diff
@@ -0,0 +1,215 @@
+            if(arch_var IN_LIST ginkgo_unsupported_archs)
+                message(FATAL_ERROR "Ginkgo does not support ${arch_var}")
+            endif()
+        else()
+            message(FATAL_ERROR
+                "CUDA ${CMAKE_CUDA_COMPILER_VERSION} does not know ${arch_var}")
+        endif()
+    else()
+        if(arch_var IN_LIST ginkgo_known_gpu_archs_name)
+            if(arch_var IN_LIST ginkgo_unsupported_archs_name)
+                message(FATAL_ERROR
+                    "Ginkgo does not support ${arch_var}")
+            endif()
+        else ()
+            message(FATAL_ERROR
+                "CUDA ${CMAKE_CUDA_COMPILER_VERSION} does not know ${arch_var}")
+        endif()
+    endif()
+endfunction()
+
+
+################################################################################
+# Function for selecting GPU arch flags for nvcc based on CUDA_ARCH_OPTION
+# Usage:
+#     ginkgo_select_nvcc_arch_flags(out_variable)
+function(ginkgo_select_nvcc_arch_flags out_variable)
+    if(CMAKE_CUDA_COMPILER_VERSION MATCHES "([0-9]+).([0-9]+).(.*)")
+        set(cuda_version ${CMAKE_MATCH_1}${CMAKE_MATCH_2})
+    endif()
+    set(ginkgo_known_gpu_archs ${ginkgo_known_gpu_archs_cuda_${cuda_version}})
+    set(ginkgo_known_gpu_archs_name
+        ${ginkgo_known_gpu_archs_name_${cuda_version}})
+    set(cuda_arch_bin_Volta ${cuda_arch_bin_Volta_${cuda_version}})
+    set(cuda_all_arch_list "")
+    foreach(__ver ${ginkgo_known_cuda_version})
+        list(APPEND cuda_all_arch_list ${ginkgo_known_gpu_archs_cuda_${__ver}})
+    endforeach()
+    list(REMOVE_DUPLICATES cuda_all_arch_list)
+
+    # set CUDA_ARCH_OPTION strings
+    set(CUDA_ARCH_OPTION "Auto" CACHE STRING
+        "Select target NVIDIA GPU achitecture.
+         Use ';' to seperate list.
+         Arch name: Kepler, Maxwell, Pascal, Volta.
+         Specified BIN, BIN(PTX), (PTX).
+         MaxPTX will select the max ptx")
+
+    set(__cuda_arch_bin "")
+    set(__cuda_arch_ptx "")
+    set(__bool_max_ptx "0")
+    foreach(__option ${CUDA_ARCH_OPTION})
+        if(__option STREQUAL "Off")
+            set(${out_variable}          ""    PARENT_SCOPE)
+            set(${out_variable}_readable "Off" PARENT_SCOPE)
+            return()
+        elseif(__option MATCHES "([0-9]+)\\(([0-9]+)\\)")
+            check_available(${CMAKE_MATCH_1})
+            check_available(${CMAKE_MATCH_2})
+            list(APPEND __cuda_arch_bin ${__option})
+        elseif(__option MATCHES "(^[0-9]+)")
+            check_available(${CMAKE_MATCH_1})
+            list(APPEND __cuda_arch_bin ${CMAKE_MATCH_1})
+        elseif(__option MATCHES "\\(([0-9]+)\\)")
+            check_available(${CMAKE_MATCH_1})
+            list(APPEND __cuda_arch_ptx ${CMAKE_MATCH_1})
+        elseif(__option STREQUAL "MaxPTX")
+            set(__bool_max_ptx "1")
+        elseif(__option STREQUAL "Auto")
+            ginkgo_detect_installed_gpus(__detect_bin)
+            check_available(${__detect_bin})
```

Comment:
Maybe I'm missing something here, but from what I understand, `ginkgo_detect_installed_gpus()` will return a list of GPU architectures if more than 1 GPU is detected, while `check_available` expects only a single architecture as input.

Shouldn't check available be called in a loop over all entries in `__detect_bin`?

Reply:
You are right. I added the foreach() to handle it now.